### PR TITLE
Add an option to tag helper providers to allow filtering

### PR DESCRIPTION
**Pull request by pranavkm**
As part of converting to source generators, one of the features we miss from MSBuild is the ability to no-op and cache
results operations if inputs are unchanged (target incrementalism). In this particular case, we're trying to
caching tag helper descriptors instead of discovering it every build.

One fairly simple solution is to cache descriptors for reference assemblies (that typically remain unchanged between compilations),
and the current assembly separately. The former can be cached and invalidated when metadatareferences change while the latter has a smaller set of types to search.

### CODE REVIEW

**Comment by pranavkm** [comment]
This comes up as part of source generation work.

File: src/Razor/Microsoft.CodeAnalysis.Razor/test/DefaultTagHelperDescriptorProviderTest.cs [hunk 1]
```diff
@@ -50,5 +50,99 @@ public void Execute_NoOpsIfCompilationIsNotSet()
...
+            Assert.Empty(context.Results.Where(f => f.GetTypeName() == testTagHelper));
+            Assert.NotEmpty(context.Results.Where(f => f.GetTypeName() == enumTagHelper));
+        }
+
+        [Fact]
+        public void Execute_WithDefaultDiscoversTagHelpersFromAssemblyAndReference()
+        {
+            // Arrange
+            var testTagHelper = "TestAssembly.TestTagHelper";
+            var enumTagHelper = "Microsoft.CodeAnalysis.Razor.Workspaces.Test.EnumTagHelper";
```
Comment (captainsafia) [inline src/Razor/Microsoft.CodeAnalysis.Razor/test/DefaultTagHelperDescriptorProviderTest.cs:121]: How does this `EnumTagHelper` come in to play? I would expect that there is a project reference to `Microsoft.CodeAnalysis.Razor.Workspaces.Test` somewhere but I am missing where that happens...

File: src/Razor/Microsoft.CodeAnalysis.Razor/src/TagHelperDiscoveryFilter.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
```
Comment (captainsafia) [inline src/Razor/Microsoft.CodeAnalysis.Razor/src/TagHelperDiscoveryFilter.cs:4]: Do we need this here?

File: src/Razor/Microsoft.CodeAnalysis.Razor/src/DefaultTagHelperDescriptorProvider.cs [hunk 3]
```diff
@@ -34,17 +34,23 @@ public void Execute(TagHelperDescriptorProviderContext context)
...
 
-            // We always visit the global namespace.
-            visitor.Visit(compilation.Assembly.GlobalNamespace);
+            if ((discoveryMode & TagHelperDiscoveryFilter.CurrentCompilation) == TagHelperDiscoveryFilter.CurrentCompilation)
+            {
+                visitor.Visit(compilation.Assembly.GlobalNamespace);
+            }
 
-            foreach (var reference in compilation.References)
+            if ((discoveryMode & TagHelperDiscoveryFilter.ReferenceAssemblies) == TagHelperDiscoveryFilter.ReferenceAssemblies)
```
Comment (captainsafia) [inline src/Razor/Microsoft.CodeAnalysis.Razor/src/DefaultTagHelperDescriptorProvider.cs:44]: Can we encapsulate these types of checks into two static methods: `ShouldDiscoverFromReferenceAssemblies` and `ShouldDiscoverFromAppAssembly` or something?

**Review by captainsafia — APPROVED** [review]
Looks legit. Left some stuff inline.

File: src/Razor/Microsoft.CodeAnalysis.Razor/src/DefaultTagHelperDescriptorProvider.cs (hunk 3 above)
Comment (pranavkm) [reply src/Razor/Microsoft.CodeAnalysis.Razor/src/DefaultTagHelperDescriptorProvider.cs:44]: If I were to create a method, it's tempting to also put the lookup inside the code which now results in an extra dictionary lookup per call. In the absence of it, it's a simple enum check, which seems like an overkill to have a method for.

File: src/Razor/Microsoft.CodeAnalysis.Razor/src/TagHelperDiscoveryFilter.cs (hunk 2 above)
Comment (pranavkm) [reply src/Razor/Microsoft.CodeAnalysis.Razor/src/TagHelperDiscoveryFilter.cs:4]: Nullable everything!

File: src/Razor/Microsoft.CodeAnalysis.Razor/test/DefaultTagHelperDescriptorProviderTest.cs (hunk 1 above)
Comment (pranavkm) [reply src/Razor/Microsoft.CodeAnalysis.Razor/test/DefaultTagHelperDescriptorProviderTest.cs:121]: The test compilation references the test project which [has these types](https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Razor/Microsoft.CodeAnalysis.Razor/test/TagHelperDescriptorFactoryTagHelpers.cs). It kinda simulates having a project reference like Mvc or Components.

**Pull request merged**